lua/lint/linters/eslint.lua

```diff
@@ -1,27 +1,74 @@
-local pattern = [[%s*(%d+):(%d+)%s+(%w+)%s+(.+%S)%s+(%S+)]]
-local groups = { 'lnum', 'col', 'severity', 'message', 'code' }
-local severity_map = {
-  ['error'] = vim.diagnostic.severity.ERROR,
-  ['warn'] = vim.diagnostic.severity.WARN,
-  ['warning'] = vim.diagnostic.severity.WARN,
+local binary_name = "eslint"
+
+local severities = {
+  vim.diagnostic.severity.WARN,
+  vim.diagnostic.severity.ERROR,
+}
+
+local diagnostic_translate_map = {
+  column = "col",
+  endColumn = "end_col",
+  endLine = "end_lnum",
+  line = "lnum",
+  message = "message",
+  ruleId = "code",
+  severity = "severity",
 }
 
 return require('lint.util').inject_cmd_exe({
   cmd = function()
-    local local_eslint = vim.fn.fnamemodify('./node_modules/.bin/eslint', ':p')
-    local stat = vim.loop.fs_stat(local_eslint)
+    local local_binary = vim.fn.fnamemodify('./node_modules/.bin/' .. binary_name, ':p')
+    local stat = vim.loop.fs_stat(local_binary)
+
     if stat then
-      return local_eslint
+      return local_binary
     end
-    return 'eslint'
+
+    return binary_name
   end,
   args = {
+    '--format',
+    'json',
     '--stdin',
     '--stdin-filename',
     function() return vim.api.nvim_buf_get_name(0) end,
   },
   stdin = true,
   stream = 'stdout',
   ignore_exitcode = true,
-  parser = require('lint.parser').from_pattern(pattern, groups, severity_map, { ['source'] = 'eslint' }),
+  parser = function(output)
+    local success, decodedData = pcall(vim.json.decode, output)
+    local json = decodedData and decodedData[1] and decodedData[1].messages or {}
+    local diagnostics = {}
+
+    if success and #json > 0 then
+      for _, json_diagnostic in ipairs(json) do
+        local diagnostic = {}
+
+        -- Translate the json diagnostic to a diagnostic.
+        for json_key, diagnostic_key in pairs(diagnostic_translate_map) do
+          if json_diagnostic[json_key] ~= vim.NIL then
+            if json_key == 'severity' then
+              diagnostic[diagnostic_key] = severities[json_diagnostic[json_key]]
+            elseif type(json_diagnostic[json_key]) == "number" then
+              diagnostic[diagnostic_key] = json_diagnostic[json_key] - 1
+            else
+              diagnostic[diagnostic_key] = json_diagnostic[json_key]
+            end
+          end
+        end
```

Comment:
We're now looking up a specific set of JSON keys, and only adding them if they aren't `vim.NIL`. This should remove all `nil` issues we've had previously.

Reply:
This is fairly similar to how `null-ls` handled this. See https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/lua/null-ls/helpers/diagnostics.lua#L184-L188.